### PR TITLE
Remove local fork of archive/tar package

### LINKED ISSUE #42402: Remove local fork of Golang's archive/tar
commit aa6a9891b09cce3d9004121294301a30d45d998d (19.03 branch) and https://github.com/moby/moby/pull/40672 (master / 20.10) re-introduced a local copy of go's archive/tar package, with a patch applied [patches/0001-archive-tar-do-not-populate-user-group-names.patch](https://github.com/moby/moby/blob/841c1f338872ca37d9fe8c699336823222dba5da/patches/0001-archive-tar-do-not-populate-user-group-names.patch).

This patch was applied for the [19.03.8 release](https://github.com/docker/docker-ce/releases/tag/v19.03.8) to improve mitigation for [CVE-2019-14271](https://nvd.nist.gov/vuln/detail/CVE-2019-14271) for some nscd configuration.

We should try to get rid of this fork again.

The discussion on https://github.com/moby/moby/pull/40672#discussion_r391306456 mentioned we should open a ticket / pull request in upstream Go to make this functionality "optional", but I think @tonistiigi also had alternatives in mind to address it.

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
Removed the local fork of `archive/tar` while retaining all existing mitigations of CVE-2019-14271. Fixes #42402

**- How I did it**
I hid the value containing the file's UID and GID from the `tar.FileInfoHeader` function [so that it would elide calling `user.LookupId`](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/archive/tar/stat_unix.go;l=28-31), analogous to the proposal in golang/go#50102.

**- How to verify it**
1. Comment out [the other mitigation](https://github.com/moby/moby/blob/559ff0ac661c1ca27bb17765629f6166d90ff499/pkg/chrootarchive/archive.go#L15-L20) and [the mitigation introduced in this PR](https://github.com/corhere/moby/blob/0ef67c6a9bdc7420210318fcef4fe0af3fc0cac6/pkg/archive/archive.go#L413-L421) and `make dynbinary`
2. Build a container image which exploits CVE-2019-14271 (an exploit payload which prints a message to stderr and exits nonzero makes it easy to spot in `dockerd`'s debug logs) and create a container from that image
3. `docker cp` from the container and observe that the exploit code has been executed
4. Uncomment the mitigation from this PR (leaving the other mitigation commented out), rebuild dockerd, and try to exploit the vulnerability again. Observe that the exploit payload does not execute.

**- Description for the changelog**
Remove local fork of `archive/tar` package

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by corhere** [comment]
I am also experimenting with having a separate binary for the chrootarchive reexecs as I described in https://github.com/moby/moby/issues/42402#issuecomment-1018068519 which has both compile-time and runtime assertions to make really, really sure that it can only be built as a static binary without cgo. I have opened a draft PR for your consideration: #43186

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -403,12 +403,49 @@ func (compression *Compression) Extension() string {
...
+		return hdr, sysStat(fi, hdr)
+	}
+	return hdr, nil
+}
+
 // FileInfoHeader creates a populated Header from fi.
-// Compared to archive pkg this function fills in more information.
-// Also, regardless of Go version, this function fills file type bits (e.g. hdr.Mode |= modeISDIR),
-// which have been deleted since Go 1.9 archive/tar.
+// Compared to the archive/tar package, this function fills in less information
```
Comment (AkihiroSuda) [inline pkg/archive/archive.go:443]: What information are lost compared to `archive/tar`?
Could you explain the comment lines in the code?

**Comment by thaJeztah** [comment]
@tonistiigi @kolyshkin @cpuguy83 PTAL

File: pkg/archive/archive.go [hunk 2]
```diff
@@ -403,12 +403,64 @@ func (compression *Compression) Extension() string {
 	return ""
 }
 
+// nosysFileInfo hides the system-dependent info of the wapped FileInfo to
```
Comment (kolyshkin) [inline pkg/archive/archive.go:406]: typo: wrapped

File: hack/vendor.sh [hunk 3]
```diff
@@ -10,18 +10,4 @@ set -x
...
-	: "${GO_VERSION:=$(awk -F '[ =]' '$1 == "ARG" && $2 == "GO_VERSION" { print $3; exit }' ./Dockerfile)}"
-	rm -rf vendor/archive/tar
-	mkdir -p vendor/archive/tar
-	echo "downloading: https://golang.org/dl/go${GO_VERSION%.0}.src.tar.gz"
-	curl -fsSL "https://golang.org/dl/go${GO_VERSION%.0}.src.tar.gz" \
-		| tar --extract --gzip --directory=vendor/archive/tar --strip-components=4 go/src/archive/tar
-	patch --strip=4 --directory=vendor/archive/tar --input="$PWD/patches/0001-archive-tar-do-not-populate-user-group-names.patch"
-fi
-
+GO111MODULE=auto go mod vendor -modfile vendor.mod
```
Comment (thaJeztah) [inline hack/vendor.sh:13]: Wah, sorry, looks like https://github.com/moby/moby/pull/43212 caused a merge conflict here, so another rebase is needed.

This now should be changed to

```bash
GO111MODULE=auto go mod tidy -modfile 'vendor.mod' -compat 1.17
GO111MODULE=auto go mod vendor -modfile vendor.mod
```


(I'll try to get more eyes on this PR to see if we can get it moving 👍)

**Review by kolyshkin — APPROVED** [review]
LGTM

In view of this, perhaps it might make sense to close https://github.com/golang/go/issues/50102 since we can achieve a similar thing without too much effort. @corhere would you do the honors to describe your solution and propose to close that one?

**Comment by corhere** [comment]
> In view of this, perhaps it might make sense to close [golang/go#50102](https://github.com/golang/go/issues/50102) since we can achieve a similar thing without too much effort.

I think the proposal should be left open. This PR implements the suggestion from https://github.com/golang/go/issues/50102#issuecomment-995070401 and reimplements the loading of (some of) the other filesystem info in userspace, so to speak. The proposed `tar.FileInfoNames` interface would reduce the amount of moby code that needs to be maintained to support new platforms, and would be beneficial to others in a similar situation who also require having atime and ctime set in their tar file headers.

**Comment by tonistiigi** [comment]
The problems with this is that it:
- Makes assumptions about the stdlib `FileInfoHeader` implementation. That the current implementation has custom code for `*tar.Header` and `statUnix` implementation in stdlib returns early before it reaches the code that can invoke the libraries that do the loading of usernames. Potentially this can change without notice in stdlib.
- We need to make sure that the new `NoLookups` function is used always. If one place forgets to call it (eg. after some code change) it breaks the security for the rest. This is also true for the proposal in https://github.com/golang/go/issues/50102

If we are ok with these risks then the implementation SGTM

**Comment by corhere** [comment]
> The problems with this is that it:
> 
> * Makes assumptions about the stdlib `FileInfoHeader` implementation. That the current implementation has custom code for `*tar.Header` and `statUnix` implementation in stdlib returns early before it reaches the code that can invoke the libraries that do the loading of usernames. Potentially this can change without notice in stdlib.

The only assumption being made is that the stdlib `tar.FileInfoHeader` implementation loads the file's username and group name iff the file's UID and GID are known to it. The `fs.FileInfo` interface cannot be widened without breaking the Go 1 Compatibility Promise so it seems highly unlikely for the mechanism used to get the UID and GID from the `FileInfo` value to change without notice. That being said, I can guard against such an occurrence by adding a unit test which asserts that `tar.FileInfoHeader(nosysFileInfo{fi}, "")` returns a `tar.Header` with zero-valued `Uname` and `Gname` fields when `fi` is a value returned from `os.Stat`.

> * We need to make sure that the new `NoLookups` function is used always. If one place forgets to call it (eg. after some code change) it breaks the security for the rest. This is also true for the proposal in [archive/tar: add FileInfoNames interface golang/go#50102](https://github.com/golang/go/issues/50102)

We need to make sure that the new function is always used in code executed inside a chroot. Or to be more precise, we need to make sure that `dlopen()` is never executed while the process is chrooted. Loading shared objects from the host's system library paths does not break security when the process's filesystem root is the host's filesystem root. And conversely, `archive/tar` is not the only package which potentially leads to `dlopen()` via glibc. Even with the existing solution of patching `archive/tar`, [security could be broken by a call to e.g. `user.Current()` or `net.Dial()` while chrooted.](https://github.com/moby/moby/issues/42402#issuecomment-1015870306) Although, as I mentioned in the PR description, [there is another layer of mitigation](https://github.com/moby/moby/blob/559ff0ac661c1ca27bb17765629f6166d90ff499/pkg/chrootarchive/archive.go#L15-L20) which prevents such oversights from actually breaking security.

**Comment by thaJeztah** [comment]
Looks like gosec linting is complaining. We had quite some "false positives" on that linter, but we should check. We updated the linter to a new version Yesterday; if `G305`  is not useful for us, we can either disable it globally or if "useful", but not for all cases, add some `//nolint: gosec` comments, e.g.

```
//nolint: gosec // Ignore G305: File traversal when extracting zip/tar archive. 
```


```
INFO [runner] linters took 1m25.662021744s with stages: goanalysis_metalinter: 1m25.57620577s
pkg/archive/archive.go:732:17: G305: File traversal when extracting zip/tar archive (gosec)
        targetPath := filepath.Join(extractDir, hdr.Linkname)
                      ^
pkg/archive/archive.go:744:17: G305: File traversal when extracting zip/tar archive (gosec)
        targetPath := filepath.Join(filepath.Dir(path), hdr.Linkname)
                      ^
pkg/archive/archive.go:1097:11: G305: File traversal when extracting zip/tar archive (gosec)
        path := filepath.Join(dest, hdr.Name)
                ^
pkg/archive/archive.go:1161:11: G305: File traversal when extracting zip/tar archive (gosec)
        path := filepath.Join(dest, hdr.Name)
                ^
pkg/archive/archive_linux.go:54:17: G305: File traversal when extracting zip/tar archive (gosec)
                Name:       filepath.Join(hdr.Name, WhiteoutOpaqueDir),
                            ^
pkg/archive/diff.go:116:11: G305: File traversal when extracting zip/tar archive (gosec)
        path := filepath.Join(dest, hdr.Name)
                ^
pkg/archive/diff.go:212:11: G305: File traversal when extracting zip/tar archive (gosec)
        path := filepath.Join(dest, hdr.Name)
                ^
pkg/tarsum/tarsum.go:249:36: G305: File traversal when extracting zip/tar archive (gosec)
            ts.currentFile = path.Join(".", path.Join("/", currentHeader.Name))
                                            ^
```

**Comment by corhere** [comment]
> Could you rebase the PR (and have a look at the linter failures? [#43185 (comment)](https://github.com/moby/moby/pull/43185#issuecomment-1036035571))

On it. Taking a peek under the hood at the gosec linter, [it indiscriminately flags any `path.Join` call with an argument derived from a `tar.Header`](https://github.com/securego/gosec/blob/7d539ed49444b2c6fd83329677ba6a9b13ecb688/rules/archive.go#L20) without checking if it is followed by a mitigation for path-traversal so lots of potential for false positives. I'll review the code to verify that all the usages are safe and [annotate them with `//#nosec` suppressions](https://github.com/securego/gosec#annotating-code) as [they are more auditable than `//nolint` suppressions](https://github.com/securego/gosec#tracking-suppressions).

> It probably doesn't hurt to have a test (just to be sure)

[Done.](https://github.com/moby/moby/pull/43185/files#diff-144ed9ddaa96c27e61b10908199e330dcd263680e7a00cebc98370f370221305R1405-R1413)

> I was actually trying to recall why we added back the vendored `pkg/archive` (after we added the other mitigation); I found [#35739 (comment)](https://github.com/moby/moby/pull/35739#issuecomment-379405517) on that topic, and some internal tickets, which may have been related to Docker Enterprise versions being on different go versions (not supporting the `osusergo` build tag) and/or FIPS-compliant builds of Docker Enterprise, but my memory is failing me.

[Thankfully it was documented at the time!](https://github.com/moby/moby/blob/61138fb5fc0f5fc1799cabfa86d28fb88a0f472c/vendor/archive/tar/README.md) Tl;dr it was a workaround required specifically for static builds on Go 1.10 with `CGO_ENABLED=1` which was made obsolete by Go 1.11 with the `osusergo` build tag. (Although in hindsight, `nosysFileInfo` might have been just as effective a workaround back then.) I'll follow up internally with the rest of the Mirantis Container Runtime team to verify that this change won't give us any trouble with FIPS-compliant builds, but otherwise it appears that the reasons for adding the back the vendored `archive/tar` have long reached their end of life.

**Comment by thaJeztah** [comment]
> [Thankfully it was documented at the time!](https://github.com/moby/moby/blob/61138fb5fc0f5fc1799cabfa86d28fb88a0f472c/vendor/archive/tar/README.md)

🤦 ah, good find! I didn't look there!

> Although in hindsight, nosysFileInfo might have been just as effective a workaround back then

If my memory serves me well, the `osusergo` option was added before the archive/tar issues existed (as it already cases grievance when using `os/user`); only before, it would happen when it was used explicitly, and now it became an issue when used implicitly (through archive/tar)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks for updating!

**Comment by thaJeztah** [comment]
@tonistiigi @cpuguy83 PTAL

**Review by AkihiroSuda — APPROVED** [review]

**Comment by thaJeztah** [comment]
I'll go ahead and bring this one in. I assume @tonistiigi is good with this (otherwise let me know)

**Pull request merged**